src/sinks/util/batch.rs

```diff
@@ -1,47 +1,93 @@
 use serde::{Deserialize, Serialize};
+use snafu::Snafu;
 use std::time::Duration;
 
-#[derive(Clone, Debug, Default, Deserialize, Serialize)]
-pub struct BatchBytesConfig {
+#[derive(Debug, Snafu)]
+pub enum BatchError {
+    #[snafu(display("Cannot configure both `max_bytes` and `max_size`"))]
+    BytesAndSize,
+    #[snafu(display("Cannot configure both `max_events` and `max_size`"))]
+    EventsAndSize,
+    #[snafu(display("This sink does not allow setting `max_bytes`"))]
+    BytesNotAllowed,
+}
+
+#[derive(Clone, Copy, Debug, Default, Deserialize, Serialize)]
+pub struct BatchConfig {
+    pub max_bytes: Option<usize>,
+    pub max_events: Option<usize>, // FIXME: Should this be "events" or "items"?
     pub max_size: Option<usize>,
     pub timeout_secs: Option<u64>,
 }
 
-impl BatchBytesConfig {
-    pub fn unwrap_or(&self, size: u64, timeout: u64) -> BatchSettings {
-        BatchSettings {
-            size: self.max_size.unwrap_or(size as usize),
-            timeout: Duration::from_secs(self.timeout_secs.unwrap_or(timeout)),
+impl BatchConfig {
+    pub fn parse_bytes(&self, bytes: u64) -> Result<usize, BatchError> {
+        match (self.max_bytes, self.max_size) {
+            (Some(_), Some(_)) => Err(BatchError::BytesAndSize),
+            (Some(bytes), None) => Ok(bytes),
+            (None, Some(size)) => Ok(size),
+            (None, None) => Ok(bytes as usize),
         }
     }
-}
 
-#[derive(Clone, Debug, Default, Deserialize, Serialize)]
-pub struct BatchEventsConfig {
-    pub max_events: Option<usize>,
-    pub timeout_secs: Option<u64>,
-}
-
-impl BatchEventsConfig {
-    pub fn unwrap_or(&self, size: u64, timeout: u64) -> BatchSettings {
-        BatchSettings {
-            size: self.max_events.unwrap_or(size as usize),
+    pub fn parse_with_bytes(
+        &self,
+        bytes: u64,
+        events: usize,
+        timeout: u64,
+    ) -> Result<BatchSettings, BatchError> {
```

Comment:
I thought about that, and like that pattern when there are many parameters. The problem is that using `BatchSettings` will then require setting timeout with `Duration::from_secs` which is extra verbose. I guess the reduced confusion will make it worth it. Any suggestions on the function naming?

Reply:
It's a bit more work, but we could consider implementing a builder-style pattern for `BatchSettings` that'd help with that. As far as naming, maybe something like `get_settings_or_default`? Don't love that but it might be a little clearer.

Reply:
I'm not convinced a builder pattern is worth it, particularly since we always specify all three parameters. The suggestion of `get_settings_or_default` is missing the size vs events differentiation for compatibility with `max_size`.

Reply:
Never mind, some sinks don't use `bytes`, so I used a builder pattern after all (with a shortcut) plus a new invention of my own for setting bytes/events from size. I hope this improves clarity.